notes/overriding.md

```diff
@@ -0,0 +1,676 @@
+
+Take two consecutive transformations:
+
+1. Set the `a` field to `1`
+2. Set the `b` field to `1`, add `1` to the `a` field
+
+```nix
+let baseRepr = self: {c = self.a + self.b;}; in
+let overlay1 = self: super: {a = 1;}; in
+let overlay2 = self: super: {b = 1; a = super.a + 1;}; in
+let applyOverlays = self:
+  let base = baseRepr self; in
+  let first = base // overlay1 self base; in
+  let second = first // overlay2 self first; in
+  second; in
+let fixpoint = applyOverlays fixpoint; in fixpoint
+```
+
+In practice, the `super // ..` and fixpoints parts can be factorised in
+dedicated helper functions.
+
+**limits**
+ - Modifications must be applied in the right order.
+ - Still need to manipulate representations.
+ - Overriding nested fields is still clumsy:
+   ```nix
+   self: super:
+   {
+     lib = (super.lib or {}) // {
+       firefoxVersion = ...;
+     };
+   }
+   ```
+
+### NixOs module system
+
+The NixOS module system takes a different approach. There, all the basic blocks
+-- the modules -- are merged together in an unspecified order. What's deciding
+the priority of one option over the other are attributes that are explicitly
+stated in the modules themselves.
+
+Nonetheless, modules need to combine different pieces of data together rather
+than just overriding one value with another. This is made difficult by the
+disappearance of the explicit `super`, compared to overlays.
+
+The issue is solved by custom merge functions, that can define how to combine
+different values of the field of a configuration. By default, when merging two
+list values, the module system only knows how to replace one value with the
+other, because there's no canonical and commutative way of merging two lists.
+However, the user can specify that these lists should in fact be concatenated,
+resulting in the possibility of defining a list of paths by pieces:
+
+```nix
+# Some module
+{
+  paths = mkOption {
+    type = types.listOf types.path;
+  };
+}
+
+# Config
+[..]
+someModule.paths = [foo/bar];
+
+# Other config
+[..]
+someModule.paths = [/bar/baz];
+
+# Resulting config: [foo/bar /bar/baz]
+```
+
+## Overriding elsewhere
+
+### CUE
+CUE allows a form of late-binding for recursive attributes:
+
+```
+$cat test.cue
+fields: {
+ a: int
+ b: a + 1
+} & {
+ a: 2
+}
+$cue eval test.cue
+fields: {
+    a: 2
+    b: 3
+}
+```
+
+Combined with [default values](https://cuelang.org/docs/tutorials/tour/types/defaults/),
+this provides an overriding mechanism:
+
+```
+$cat test2.cue
+fields: {
+ a: int | *1
+ b: a + 1
+} & {
+ a: 2
+}
+$cue eval test2.cue
+fields: {
+    a: 2
+    b: 3
+}
+```
+
+The basics of this overriding mechanism are close in spirit to what we want to
+achieve in this proposal. This is similar to Nickel's current `default`
+behavior, but with the desired late-bound merging. This is however not enough
+for Nix use cases: we will need to add other features to the merge system.
```

Comment:
What other features?

Cf what I said the other day about naming stuff. You can start by naming the features that you need, and then explain which features are met or not by different solutions. The text will flow much better. And it will help frame the conversation and the documentation in the future.